File: examples/python-local/transport_adapters.py
"""Dependency-free auth/retry and idempotency adapter examples."""

from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path
import tempfile
from typing import Any

from googlechatai import FileIdempotencyStore, request_json_with_retry


class ExampleTokenBroker:
    def __init__(self) -> None:
        self._tokens = {
            "user:adam@example.com": {
                "stale": "stale-user-token",
                "fresh": "fresh-user-token",
            }
        }

    def get_access_token(
        self,
        *,
        principal_id: str,
        force_refresh: bool = False,
    ) -> dict[str, Any]:
        token = self._tokens.get(principal_id)
        if token is None:
            raise RuntimeError(f"No token lease is available for {principal_id}.")
        return {
            "access_token": token["fresh"] if force_refresh else token["stale"],
            "refreshed": force_refresh,
        }


@dataclass
class StoredClaim:
    first_seen_at_ms: int
    last_seen_at_ms: int
    expires_at_ms: int
    seen_count: int
    metadata: dict[str, Any] | None = None


class CompareAndSetBackend:
    def __init__(self) -> None:
        self._records: dict[str, StoredClaim] = {}

    def create_if_absent(self, key: str, value: StoredClaim) -> bool:
        if key in self._records:
            return False
        self._records[key] = value
        return True

    def get(self, key: str) -> StoredClaim | None:
        return self._records.get(key)

    def update_seen(self, key: str, now_ms: int) -> StoredClaim | None:
        current = self._records.get(key)
        if current is None:
            return None
        current.last_seen_at_ms = now_ms
        current.seen_count += 1
        return current

    def replace(self, key: str, value: StoredClaim) -> None:
        self._records[key] = value


class ExternalIdempotencyStore:
    def __init__(
        self,
        *,
        backend: CompareAndSetBackend,
        default_ttl_ms: int = 10 * 60 * 1000,
    ) -> None:
        self.backend = backend
        self.default_ttl_ms = default_ttl_ms

    def claim(
        self,
        key: str,
        *,
        ttl_ms: int | None = None,
        now_ms: int,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        existing = self.backend.get(key)
        expired = existing is not None and existing.expires_at_ms <= now_ms
        if existing is not None and not expired:
            updated = self.backend.update_seen(key, now_ms)
            return claim_from_record(key, updated, claimed=False, duplicate=True)

        effective_ttl_ms = ttl_ms or self.default_ttl_ms
        record = StoredClaim(
            first_seen_at_ms=now_ms,
            last_seen_at_ms=now_ms,
            expires_at_ms=now_ms + effective_ttl_ms,
            seen_count=1,
            metadata=metadata,
        )
        if expired:
            self.backend.replace(key, record)
            return claim_from_record(key, record, claimed=True, duplicate=False)

        inserted = self.backend.create_if_absent(key, record)
        if not inserted:
            updated = self.backend.update_seen(key, now_ms)
            return claim_from_record(key, updated, claimed=False, duplicate=True)
        return claim_from_record(key, record, claimed=True, duplicate=False)


def iso(ms: int) -> str:
    from datetime import datetime, timezone

    return datetime.fromtimestamp(ms / 1000, tz=timezone.utc).isoformat().replace(
        "+00:00",
        "Z",
    )


def claim_from_record(
    key: str,
    record: StoredClaim | None,
    *,
    claimed: bool,
    duplicate: bool,
) -> dict[str, Any]:
    if record is None:
        raise RuntimeError("Expected stored idempotency claim.")
    return {
        "key": key,
        "claimed": claimed,
        "duplicate": duplicate,
        "firstSeenAt": iso(record.first_seen_at_ms),
        "lastSeenAt": iso(record.last_seen_at_ms),
        "expiresAt": iso(record.expires_at_ms),
        "seenCount": record.seen_count,
        "metadata": record.metadata,
    }


def demo_retrying_user_read() -> dict[str, Any]:
    token_broker = ExampleTokenBroker()
    authorizations: list[str] = []

    def get_access_token(*, force_refresh: bool = False) -> dict[str, Any]:
        return token_broker.get_access_token(
            principal_id="user:adam@example.com",
            force_refresh=force_refresh,
        )

    def send(request: dict[str, Any]) -> dict[str, Any]:
        authorizations.append(request["headers"]["authorization"])
        if len(authorizations) == 1:
            return {
                "ok": False,
                "status": 401,
                "json": {"error": {"status": "UNAUTHENTICATED"}},
                "headers": {},
            }
        return {
            "ok": True,
            "status": 200,
            "json": {"spaces": []},
            "headers": {},
        }

    result = request_json_with_retry(
        method="GET",
        url="https://chat.googleapis.com/v1/spaces?pageSize=1",
        principal="user",
        get_access_token=get_access_token,
        send=send,
        sleep=lambda _delay_ms: None,
    )

    return {
        "ok": result.ok,
        "status": result.status,
        "attempts": result.attempts,
        "refreshed": result.refreshed,
        "replayedAfter401": result.replayed_after_401,
        "rawTokensPrinted": False,
    }


def demo_idempotency_stores() -> dict[str, Any]:
    with tempfile.TemporaryDirectory() as tmpdir:
        local_store = FileIdempotencyStore(
            Path(tmpdir) / "idempotency.json",
            default_ttl_ms=60_000,
        )
        external_store = ExternalIdempotencyStore(
            backend=CompareAndSetBackend(),
            default_ttl_ms=60_000,
        )

        local_first = local_store.claim("event-id-hash-local", now_ms=1_000)
        local_duplicate = local_store.claim("event-id-hash-local", now_ms=2_000)
        local_after_expiry = local_store.claim(
            "event-id-hash-local",
            now_ms=62_000,
        )
        external_first = external_store.claim(
            "event-id-hash-external",
            now_ms=1_000,
            metadata={"source": "direct_chat_event"},
        )
        external_duplicate = external_store.claim(
            "event-id-hash-external",
            now_ms=2_000,
        )
        external_after_expiry = external_store.claim(
            "event-id-hash-external",
            now_ms=62_000,
        )

        return {
            "localFileStore": {
                "firstClaimed": local_first.claimed,
                "duplicateSuppressed": local_duplicate.duplicate,
                "seenCount": local_duplicate.seen_count,
                "afterExpiryClaimed": local_after_expiry.claimed,
            },
            "externalCompareAndSetStore": {
                "firstClaimed": external_first["claimed"],
                "duplicateSuppressed": external_duplicate["duplicate"],
                "seenCount": external_duplicate["seenCount"],
                "afterExpiryClaimed": external_after_expiry["claimed"],
            },
        }


if __name__ == "__main__":
    print(
        json.dumps(
            {
                "retryingUserRead": demo_retrying_user_read(),
                "idempotency": demo_idempotency_stores(),
            },
            indent=2,
        )
    )

File: examples/node-local-runtime/transport-adapters.mjs
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";

import {
  FileIdempotencyStore,
  requestJsonWithRetry,
} from "../../packages/node/dist/index.js";

class ExampleTokenBroker {
  #tokens = new Map();

  constructor() {
    this.#tokens.set("user:adam@example.com", {
      stale: "stale-user-token",
      fresh: "fresh-user-token",
    });
  }

  async getAccessToken({ principalId, forceRefresh = false }) {
    const token = this.#tokens.get(principalId);
    if (!token) {
      throw new Error(`No token lease is available for ${principalId}.`);
    }

    return {
      accessToken: forceRefresh ? token.fresh : token.stale,
      refreshed: forceRefresh,
    };
  }
}

class CompareAndSetBackend {
  #records = new Map();

  async createIfAbsent(key, value) {
    if (this.#records.has(key)) {
      return false;
    }
    this.#records.set(key, value);
    return true;
  }

  async get(key) {
    return this.#records.get(key) ?? null;
  }

  async update(key, updateRecord) {
    const current = this.#records.get(key);
    if (!current) {
      return null;
    }
    const next = updateRecord(current);
    this.#records.set(key, next);
    return next;
  }

  async replace(key, value) {
    this.#records.set(key, value);
  }
}

class ExternalIdempotencyStore {
  constructor({ backend, defaultTtlMs = 10 * 60 * 1000 }) {
    this.backend = backend;
    this.defaultTtlMs = defaultTtlMs;
  }

  async claim({ key, ttlMs = this.defaultTtlMs, nowMs = Date.now(), metadata }) {
    const existing = await this.backend.get(key);
    const expired = existing && existing.expiresAtMs <= nowMs;

    if (existing && !expired) {
      const updated = await this.backend.update(key, (record) => ({
        ...record,
        lastSeenAtMs: nowMs,
        seenCount: record.seenCount + 1,
      }));
      return claimFromRecord(key, updated, {
        claimed: false,
        duplicate: true,
      });
    }

    const record = {
      firstSeenAtMs: nowMs,
      lastSeenAtMs: nowMs,
      expiresAtMs: nowMs + ttlMs,
      seenCount: 1,
      metadata,
    };

    if (expired) {
      await this.backend.replace(key, record);
      return claimFromRecord(key, record, {
        claimed: true,
        duplicate: false,
      });
    }

    const inserted = await this.backend.createIfAbsent(key, record);

    if (!inserted) {
      const updated = await this.backend.update(key, (current) => ({
        ...current,
        lastSeenAtMs: nowMs,
        seenCount: current.seenCount + 1,
      }));
      return claimFromRecord(key, updated, {
        claimed: false,
        duplicate: true,
      });
    }

    return claimFromRecord(key, record, {
      claimed: true,
      duplicate: false,
    });
  }
}

function claimFromRecord(key, record, { claimed, duplicate }) {
  return {
    key,
    claimed,
    duplicate,
    firstSeenAt: new Date(record.firstSeenAtMs).toISOString(),
    lastSeenAt: new Date(record.lastSeenAtMs).toISOString(),
    expiresAt: new Date(record.expiresAtMs).toISOString(),
    seenCount: record.seenCount,
    metadata: record.metadata ?? null,
  };
}

async function demoRetryingUserRead() {
  const tokenBroker = new ExampleTokenBroker();
  const authHeaders = [];

  const result = await requestJsonWithRetry(
    {
      url: "https://chat.googleapis.com/v1/spaces?pageSize=1",
      method: "GET",
      principal: "user",
    },
    {
      getAccessToken: ({ forceRefresh }) =>
        tokenBroker.getAccessToken({
          principalId: "user:adam@example.com",
          forceRefresh,
        }),
      fetch: async (_url, init) => {
        authHeaders.push(init.headers.authorization);

        if (authHeaders.length === 1) {
          return new Response(
            JSON.stringify({ error: { status: "UNAUTHENTICATED" } }),
            { status: 401, headers: { "content-type": "application/json" } },
          );
        }

        return new Response(JSON.stringify({ spaces: [] }), {
          status: 200,
          headers: { "content-type": "application/json" },
        });
      },
      sleepMs: async () => {},
    },
  );

  return {
    ok: result.ok,
    status: result.status,
    attempts: result.attempts,
    refreshed: result.refreshed,
    replayedAfter401: result.replayedAfter401,
    rawTokensPrinted: false,
  };
}

async function demoIdempotencyStores() {
  const tempDir = await fs.mkdtemp(
    path.join(os.tmpdir(), "googlechatai-sdk-example-"),
  );
  const localStore = new FileIdempotencyStore({
    filePath: path.join(tempDir, "idempotency.json"),
    defaultTtlMs: 60_000,
  });
  const externalStore = new ExternalIdempotencyStore({
    backend: new CompareAndSetBackend(),
    defaultTtlMs: 60_000,
  });

  try {
    const localFirst = await localStore.claim({
      key: "event-id-hash-local",
      nowMs: 1_000,
    });
    const localDuplicate = await localStore.claim({
      key: "event-id-hash-local",
      nowMs: 2_000,
    });
    const localAfterExpiry = await localStore.claim({
      key: "event-id-hash-local",
      nowMs: 62_000,
    });
    const externalFirst = await externalStore.claim({
      key: "event-id-hash-external",
      nowMs: 1_000,
      metadata: { source: "direct_chat_event" },
    });
    const externalDuplicate = await externalStore.claim({
      key: "event-id-hash-external",
      nowMs: 2_000,
    });
    const externalAfterExpiry = await externalStore.claim({
      key: "event-id-hash-external",
      nowMs: 62_000,
    });

    return {
      localFileStore: {
        firstClaimed: localFirst.claimed,
        duplicateSuppressed: localDuplicate.duplicate,
        seenCount: localDuplicate.seenCount,
        afterExpiryClaimed: localAfterExpiry.claimed,
      },
      externalCompareAndSetStore: {
        firstClaimed: externalFirst.claimed,
        duplicateSuppressed: externalDuplicate.duplicate,
        seenCount: externalDuplicate.seenCount,
        afterExpiryClaimed: externalAfterExpiry.claimed,
      },
    };
  } finally {
    await fs.rm(tempDir, { recursive: true, force: true });
  }
}

const output = {
  retryingUserRead: await demoRetryingUserRead(),
  idempotency: await demoIdempotencyStores(),
};

console.log(JSON.stringify(output, null, 2));
